### ISSUE TITLE
ENH: multi dimensional wasserstein/earth mover distance in Scipy

### ISSUE THREAD

**Issue by com3dian** [scipy/scipy#17290]
### Is your feature request related to a problem? Please describe.

Currently, Scipy has its own implementation of the wasserstein distance ->  [`scipy.stats.wasserstein_distance`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.wasserstein_distance.html). However, the [scipy.stats.wasserstein_distance](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.wasserstein_distance.html) function only works with one dimensional data. @[Eight1911](https://github.com/Eight1911) created an issue [#10382](https://github.com/scipy/scipy/issues/10382) in 2019 suggesting a more general support for multi-dimensional data. As far as I know, his pull request was not merged and the link to his script is now invalid.

*A bit background below* 

The wasserstein distance, also called the Earth mover distance or the optimal transport distance, is defined as a similarity metric between two probability distribution. In the discrete case, the wasserstein distance can be understood as the cost of an optimal transport plan between two samples. A brief and intuitive introduction can be found [here](https://lilianweng.github.io/posts/2017-08-20-gan/#what-is-wasserstein-distance). 

After it was firstly introduced in the Monge problem, the wasserstein distance has been studied for many years. It has been widely used in many areas to compare discrete distributions. For example, it was used to compare color histograms in computer vision, or as a similarity metric for anomaly detection. In the WGAN  neural network framework, it was used as a loss function.

In one dimensional case, the wassrstein distance is equal to the energy distance. However, this does not hold for multi-dimensional metric space.

### Describe the solution you'd like.

I'm suggesting Scipy to pick up that issue and carry on adding a multi-dimensional wasserstein distance in Scipy. Also, I'm willing to introduce it into Scipy if the Scipy developers are happy with that. My current idea is to implement a function that supports multiple 'backend' techniques. The function would be like this:

```python
wass_nd(u, v, u_weight = None, v_weight = None, method = ['linear_programming', 'network_simplex', 'sinkhorn', ...])
```

and return a list of wasserstein values. Each element in the list corresponds to an element in the argument `method`.

For now I'm considering that the function should support at least the linear programming method, as it is a well studied approach in the optimal transport theory. I wrote a python script for that based on [Vincent Herrmann's blog](https://vincentherrmann.github.io/blog/wasserstein/) using `scipy.optimize.linprog` and tested it against [`scipy.stats.wasserstein_distance`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.wasserstein_distance.html) on one dimensional data. Despite the comparatively slow speed, the [results](https://github.com/com3dian/wass_nd/tree/main) show consistency. 

The Sinkhorn algorithm is another approach for the optimal transport problem, firstly introduced in [this paper](https://papers.nips.cc/paper/2013/file/af21d0c97db2e27e13572cbf59eb343d-Paper.pdf). The Sinkhorn approach provides an approximate technique for the optimal transport problem by adding an entropy regularization to the original optimal transport.

More details about the LP and Sinkhorn approaches can be found in [Computational Optimal Transport](https://arxiv.org/abs/1803.00567) by [Gabriel Peyré](https://arxiv.org/search/stat?searchtype=author&query=Peyr%C3%A9%2C+G) and [Marco Cuturi](https://arxiv.org/search/stat?searchtype=author&query=Cuturi%2C+M). I'm not so familiar with network simplex algorithms but there are also many methods among them used to solve the optimal transport problem, like the revised simplex algorithm and the shortsimplex algorithm.

### Describe alternatives you've considered.

In python, there are two alternatives for wasserstein distance. One is the wasserstein distance in the `POT` package, the other is the `Cv2.EMD` method. However, both of them have some issues when the sample size goes too large.

**Comment by com3dian** [comment]
#17201 
cc: @rgommers @morales-gregorio @Eight1911

**Comment by mdhaber** [comment]
I develop scientific Python more than I use it, but this is one of the few features I have actually needed for research. Still, I lean toward suggesting, @com3dian, that you propose the enhancement to POT. They would need to update their minimum supported SciPy version to 1.6 because that's when `linprog` `method='HiGHS'` was introduced, but if you can show the superiority of a simple linear programming solution for very large problems, perhaps they would work with you to add that method and go further. 
There are a few reasons I feel this way, but the best ones are probably that it would would likely be easier to find highly qualified reviewers over there (perhaps I'm wrong - other maintainers are welcome to chime in), and POT seems to be _the_ Python library for this sort of thing (IMO), whereas SciPy is a more general-purpose library. I can maybe see extending the existing `wassertstein_distance` to multiple dimensions using `linprog`, but I wouldn't want to get `scipy.stats` into the business of a many-`method`, fully-featured EMD suite.

What do you think @tupui @tirthasheshpatel @chrisb83 @ev-br?

**Comment by tupui** [comment]
No my area so I would trust your judgment Matt. From the previous discussion, it also seemed that it could find it's place in distance instead.

**Comment by com3dian** [comment]
Thank you for replying @mdhaber ! A mini-expansion with `linprog` sounds cool to me. And do you think working on the current function or developing a new function probably named `wasserstein_nd_distance` would be a healthier way?

Also regarding @tupui 's comment, `scipy.stats` already has the `wasserstein_distance` for long so it becomes weird to me if a new function solving the same problem would show up in a different place. So I'm suggesting either moving both to distance or keeping it in stats.

**Comment by tupui** [comment]
If from a stats perspective it makes less sense than a distance perspective, we can make the move. But again, I was just reading up the linked issue and this is not my territory 😃

**Comment by mdhaber** [comment]
> And do you think working on the current function or developing a new function probably

I thought about that, and I'd prefer that it keep the same function name if the ND functionally fits naturally into the existing signature. What would your ideal signature be?

**Comment by com3dian** [comment]
> What would your ideal signature be?
The current args are sufficient and necessary for an ND function, a new function doesn't have to take anymore args.

**Comment by mdhaber** [comment]
Then I think it's reasonable to just extend the existing function, if the plan is to write a fairly simple implementation with `linprog` (or, I'd suggest `milp`, really).

**Comment by com3dian** [comment]
Extending the current function sounds cool to me(and moving to distance as well I think?). 
However, I don't see any reason to apply `milp` instead of `linprog`, is that simply an integral constrained version for `linprog` or does it also has some other necessary feature?

**Comment by mdhaber** [comment]
There are two small reasons: 
* If we ever need to debug, it will be easier slightly easier to work with `milp`
* `milp` accepts the arguments in the form that HiGHS wants them, so no conversion from equality and upper bound constraints to the HIGHS form is needed

So if there are no reasons not to, I lean toward MILP. Under the hood, both just call `highs_wrapper`, and `I think `high_wrapper` will do the same thing coming from `linprog` or `milp` if no `solver` is provided.

**Comment by com3dian** [comment]
All right, then I will replace `linprog` with `milp` and start a PR soon, would you like to review that @mdhaber?

**Comment by mdhaber** [comment]
I just ran across gh-7804, which concluded that EMD should be added in `scipy.spatial.distance`. 
I also realized that if we were to add this to the existing `wasserstein_distance`, it might not be easy to add an `axis` argument later.
Finally, gh-14540 (network simplex) seems to be getting closer to merging, and that might just be superior to LP for this purpose.
Let me think about this a little more today and get back to you. In the meantime, would you be interested in trying to use gh-14540 to implement this? (Or, since that came from the NetworkX implementation, you could use the network simplex algorithm there.) If it's also exact and clearly faster, there would be no need for an LP implementation, right?

**Comment by com3dian** [comment]
It's okay for me to create a new function for EMD if `wasserstein_distance` does not fit in the signature. However, I'm not sure what `axis` argument you are talking about(If you mean the `axis` argument in `[sklearn.metrics](https://scikit-learn.org/stable/modules/classes.html#module-sklearn.metrics).pairwise_distances_argmin` it is different with EMD as it says that the pairwise_distance_argmin 'computes minimum distances between one point and a set of points' while EMD returns the minimum cost of transport between two datasets).
I can try with the network simplex approach definitely, but I suggest that I start a PR first containing only the LP implementation and wasserstein tests then update the function with the network simplex algorithm in the next interaction once it is merged. 
> you could use the network simplex algorithm there
Do you mean I can call networkx functions?

Also it seems now the progress is becoming a bit slow as the content is expanding and I probably won't be able to have enough time to work on this in December and January. Would you like to have a quick call next week to explain a bit of your ideas? So that we can have solid plan for this.

**Comment by mdhaber** [comment]
Currently, `wasserstein_distance` works only on 1D arrays, but it is natural to extend it to accept multidimensional arrays and operate along all rows (or columns, or other slices) independently but in parallel. `axis` is used throughout `stats` to tell SciPy which axis of a multidimensional array it should take the statistic over.  This is a competing way to interpret multidimensional input (many 1D inputs).

Oops, no, I didn't mean we can call NetworkX in SciPy. We would call the network simplex algorithm in SciPy once that PR merges. I just meant that the NetworkX version might be easier to try since that PR is not yet merged.

Go ahead and open a PR extending `wasserstein_distance` using LP.  Don't worry about my recent comments - if the interface needs to change, I'll take care of it, since I can't seem to come to a decision. It is more important to let you get the code in, and there is much more to review beside the interface. Let's see how it goes, and if we need a call, I'm free all next week.

**Comment by com3dian** [comment]
Hi I just started a PR containing extended `wassserstein_distance` with LP and the updated `TestWassersteinDistance` class(https://github.com/scipy/scipy/pull/17473). However there are several failing test and as a noob at this I don't have much clue, can you maybe give me some hint how to fix that?
Now the pull request only have `stats.wasserstein_distance` and its test, `spatial.EMD` can be add later on simply calling it.

By the way, I also wrote a script with `networkx.network_simplex`and tested that. Yet it seems to me not very useful since `network_simplex` only promises a solution if the edge weight(equivalent to distance in optimal the transport problem) has only integer values.

**Comment by mdhaber** [comment]
Great. Added some comments there to help you debug.  Yeah, we should not add `spatial.EMD` in this PR; that would be a separate thing, and someone else would need to review that other than me.
